### PR TITLE
Tc2 local cache

### PR DESCRIPTION
**Pull request by hernandezvictoria**
## Description

- created local caching system
- using hm for cache and priority queue with custom priority computation for determining which products to flush

## Milestones
- TC 2

## Resources
- https://www.bing.com/videos/riverview/relatedvideo?&q=how+to+implement+custom+caching+system+with+postgresql+node+js&&mid=4871CF3734B34267C5944871CF3734B34267C594&&mcid=3FEE40CD65E046C5AF9FEE338EC96FF0&FORM=VRDGAR — helpful video that explained caching strategies well
- https://medium.com/building-the-system/how-to-store-dates-and-times-in-postgresql-269bda8d6403 — time in PostgreSQL
- https://www.internalfb.com/metamate/24397505419910362 time comparison PostgreSQL vs. NodeJS
- https://www.geeksforgeeks.org/javascript/difference-between-new-date-and-datenow-in-javascript/ using new date() vs date.now()
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map — map methods
- https://www.npmjs.com/package/@datastructures-js/priority-queue#fromarray --- priority queue documentation

## Test Plan
- see test-cache.js for some test cases, I played around with numbers and functions tested so not all edge cases showed
- mostly verified functionality of helper functions and main functions

### CODE REVIEW

File: backend/routes/local-cache.js [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        }
+        return a.priority - b.priority;
+    });
+    productImageCache = new Map(); // productId -> image url, timestamp
+}
+
+// returns a number representing the priority of the product (based on how used it is)
+const computeInitialPriority = async (productId) => {
+    //higher click velocity -> higher priority score, less likely to be flushed from cache
+    if (productImageCache.has(productId)) {
```
Comment (teaenshin) [inline backend/routes/local-cache.js:33]: when do we computer initial priority when the productID is in the cache? Is it only when we are replacing an expired entry? Then wouldn't `currentTime - productTime >= TTL` always be true and cause this to return lowest pri?

File: backend/routes/local-cache.js [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+let productQueue;
+let productImageCache;
+
+const createQueueAndCache = () => {
+    // priority queue for products, less interacted with products have higher priority (first to be flushed from cache)
+    productQueue = new PriorityQueue((a, b) => {
+        if (a.priority === b.priority) { // if they are tied, sort by recency
+            const aTime = productImageCache.has(a.productId) ? productImageCache.get(a.productId).timestamp.getTime() : Date.now();
+            const bTime = productImageCache.has(b.productId) ? productImageCache.get(b.productId).timestamp.getTime() : Date.now();
```
Comment (emilyzwang) [inline backend/routes/local-cache.js:22]: is there a case where the product would be in the priority queue but not the cache?

File: backend/routes/local-cache.js (hunk 1 above)
Comment (hernandezvictoria) [reply backend/routes/local-cache.js:33]: Yeah it's only when it is initially loaded into the cache. I think it is actually the opposite. Since it only gets computed when products are first added to the cache, the productTime will be really close to currentTime because the product gets added to the cache immediately before computing its priority. Thus, currentTime - productTime should never be greater than TTL (1 day)
